File: src-tauri/src/commands/diff.rs
use tauri::{command, State};
use crate::commands::state::AppState;
use crate::error::{AppError, AppResult};
use crate::git::{
    blame::{
        get_file_blame as git_get_file_blame, get_file_history as git_get_file_history,
        BlameHunkItem, FileHistoryItem,
    },
    diff::{
        get_commit_file_diff as git_get_commit_file_diff, get_working_tree_file_diff,
        stage_all as git_stage_all, stage_file as git_stage_file,
        stage_hunk as git_stage_hunk, unstage_all as git_unstage_all, unstage_file as git_unstage_file,
        unstage_hunk as git_unstage_hunk, FileDiffDetail,
    },
    ignore::{add_pattern_to_gitignore, generate_smart_gitignore},
    repo::open_repository as git_open_repo,
    safety::{
        delete_trash_snapshot as git_delete_trash, discard_all_changes as git_discard_all,
        discard_file_changes as git_discard_file, list_trash_snapshots as git_list_trash,
        restore_trash_snapshot as git_restore_trash,
    },
    status::{get_working_tree_status, WorkingTreeStatus},
};
use crate::storage::trash::TrashSnapshotItem;

#[command]
pub async fn get_status(path: String) -> AppResult<WorkingTreeStatus> {
    tokio::task::spawn_blocking(move || {
        let repo = git_open_repo(&path)?;
        get_working_tree_status(&repo)
    })
    .await
    .map_err(|e| AppError::Internal(e.to_string()))?
}

#[command]
pub async fn get_file_diff(
    path: String,
    file_path: String,
    staged: bool,
    ignore_whitespace: Option<bool>,
) -> AppResult<FileDiffDetail> {
    tokio::task::spawn_blocking(move || {
        let repo = git_open_repo(&path)?;
        get_working_tree_file_diff(&repo, &file_path, staged, ignore_whitespace)
    })
    .await
    .map_err(|e| AppError::Internal(e.to_string()))?
}

#[command]
pub async fn get_commit_file_diff(
    path: String,
    commit_id: String,
    file_path: String,
    ignore_whitespace: Option<bool>,
) -> AppResult<FileDiffDetail> {
    tokio::task::spawn_blocking(move || {
        let repo = git_open_repo(&path)?;
        git_get_commit_file_diff(&repo, &commit_id, &file_path, ignore_whitespace)
    })
    .await
    .map_err(|e| AppError::Internal(e.to_string()))?
}

#[command]
pub async fn stage_file(path: String, file_path: String) -> AppResult<()> {
    tokio::task::spawn_blocking(move || {
        let repo = git_open_repo(&path)?;
        git_stage_file(&repo, &file_path)
    })
    .await
    .map_err(|e| AppError::Internal(e.to_string()))?
}

#[command]
pub async fn unstage_file(path: String, file_path: String) -> AppResult<()> {
    tokio::task::spawn_blocking(move || {
        let repo = git_open_repo(&path)?;
        git_unstage_file(&repo, &file_path)
    })
    .await
    .map_err(|e| AppError::Internal(e.to_string()))?
}

#[command]
pub async fn stage_all(path: String) -> AppResult<()> {
    tokio::task::spawn_blocking(move || {
        let repo = git_open_repo(&path)?;
        git_stage_all(&repo)
    })
    .await
    .map_err(|e| AppError::Internal(e.to_string()))?
}

#[command]
pub async fn unstage_all(path: String) -> AppResult<()> {
    tokio::task::spawn_blocking(move || {
        let repo = git_open_repo(&path)?;
        git_unstage_all(&repo)
    })
    .await
    .map_err(|e| AppError::Internal(e.to_string()))?
}

#[command]
pub async fn stage_hunk(path: String, file_path: String, hunk_index: usize) -> AppResult<()> {
    tokio::task::spawn_blocking(move || {
        let repo = git_open_repo(&path)?;
        git_stage_hunk(&repo, &file_path, hunk_index)
    })
    .await
    .map_err(|e| AppError::Internal(e.to_string()))?
}

#[command]
pub async fn unstage_hunk(path: String, file_path: String, hunk_index: usize) -> AppResult<()> {
    tokio::task::spawn_blocking(move || {
        let repo = git_open_repo(&path)?;
        git_unstage_hunk(&repo, &file_path, hunk_index)
    })
    .await
    .map_err(|e| AppError::Internal(e.to_string()))?
}

#[command]
pub async fn discard_file_changes(
    path: String,
    file_path: String,
    state: State<'_, AppState>,
) -> AppResult<()> {
    let store = state.trash_store.clone();
    tokio::task::spawn_blocking(move || {
        let repo = git_open_repo(&path)?;
        git_discard_file(&repo, &store, &file_path)
    })
    .await
    .map_err(|e| AppError::Internal(e.to_string()))?
}

#[command]
pub async fn discard_all_changes(
    path: String,
    state: State<'_, AppState>,
) -> AppResult<usize> {
    let store = state.trash_store.clone();
    tokio::task::spawn_blocking(move || {
        let repo = git_open_repo(&path)?;
        git_discard_all(&repo, &store)
    })
    .await
    .map_err(|e| AppError::Internal(e.to_string()))?
}

#[command]
pub async fn list_trash_snapshots(
    path: String,
    state: State<'_, AppState>,
) -> AppResult<Vec<TrashSnapshotItem>> {
    let store = state.trash_store.clone();
    tokio::task::spawn_blocking(move || {
        git_list_trash(&store, &path)
    })
    .await
    .map_err(|e| AppError::Internal(e.to_string()))?
}

#[command]
pub async fn restore_trash_snapshot(
    path: String,
    snapshot_id: i64,
    state: State<'_, AppState>,
) -> AppResult<()> {
    let store = state.trash_store.clone();
    tokio::task::spawn_blocking(move || {
        let repo = git_open_repo(&path)?;
        git_restore_trash(&repo, &store, snapshot_id)
    })
    .await
    .map_err(|e| AppError::Internal(e.to_string()))?
}

#[command]
pub async fn delete_trash_snapshot(
    snapshot_id: i64,
    state: State<'_, AppState>,
) -> AppResult<()> {
    let store = state.trash_store.clone();
    tokio::task::spawn_blocking(move || {
        git_delete_trash(&store, snapshot_id)
    })
    .await
    .map_err(|e| AppError::Internal(e.to_string()))?
}

#[command]
pub async fn get_file_blame(
    path: String,
    file_path: String,
) -> AppResult<Vec<BlameHunkItem>> {
    tokio::task::spawn_blocking(move || {
        let repo = git_open_repo(&path)?;
        git_get_file_blame(&repo, &file_path)
    })
    .await
    .map_err(|e| AppError::Internal(e.to_string()))?
}

#[command]
pub async fn get_file_history(
    path: String,
    file_path: String,
    limit: Option<usize>,
) -> AppResult<Vec<FileHistoryItem>> {
    tokio::task::spawn_blocking(move || {
        let repo = git_open_repo(&path)?;
        git_get_file_history(&repo, &file_path, limit.unwrap_or(50))
    })
    .await
    .map_err(|e| AppError::Internal(e.to_string()))?
}

#[command]
pub async fn add_to_gitignore(path: String, pattern: String) -> AppResult<()> {
    tokio::task::spawn_blocking(move || {
        let repo = git_open_repo(&path)?;
        let workdir = repo.workdir().ok_or_else(|| AppError::Internal("Bare repo has no workdir".into()))?;
        add_pattern_to_gitignore(workdir.to_str().unwrap_or(&path), &pattern)
    })
    .await
    .map_err(|e| AppError::Internal(e.to_string()))?
}

#[command]
pub async fn generate_standard_gitignore(path: String) -> AppResult<Vec<String>> {
    tokio::task::spawn_blocking(move || {
        let repo = git_open_repo(&path)?;
        let workdir = repo.workdir().ok_or_else(|| AppError::Internal("Bare repo has no workdir".into()))?;
        generate_smart_gitignore(workdir.to_str().unwrap_or(&path))
    })
    .await
    .map_err(|e| AppError::Internal(e.to_string()))?
}

File: src/lib/api/diff.ts
import type {
  BlameHunkItem,
  FileDiffDetail,
  FileHistoryItem,
  TrashSnapshotItem,
  WorkingTreeStatus,
} from '../types';
import { isTauri, invokeWithTimeout } from './client';
import { getMockFileDiff, getMockTrashSnapshots, getMockWorkingTreeStatus } from './mocks';

export async function getWorkingTreeStatus(path: string): Promise<WorkingTreeStatus> {
  if (isTauri) {
    return await invokeWithTimeout<WorkingTreeStatus>('get_status', { path }, 15000);
  }
  return getMockWorkingTreeStatus();
}

export async function getFileDiff(
  path: string,
  filePath: string,
  staged: boolean,
  ignoreWhitespace?: boolean
): Promise<FileDiffDetail> {
  if (isTauri) {
    return await invokeWithTimeout<FileDiffDetail>(
      'get_file_diff',
      { path, filePath, staged, ignoreWhitespace },
      15000
    );
  }
  return getMockFileDiff(filePath, staged);
}

export async function getCommitFileDiff(
  path: string,
  commitId: string,
  filePath: string,
  ignoreWhitespace?: boolean
): Promise<FileDiffDetail> {
  if (isTauri) {
    return await invokeWithTimeout<FileDiffDetail>(
      'get_commit_file_diff',
      { path, commitId, filePath, ignoreWhitespace },
      15000
    );
  }
  return getMockFileDiff(filePath, false);
}

export async function stageFile(path: string, filePath: string): Promise<void> {
  if (isTauri) {
    await invokeWithTimeout('stage_file', { path, filePath }, 15000);
  }
}

export async function unstageFile(path: string, filePath: string): Promise<void> {
  if (isTauri) {
    await invokeWithTimeout('unstage_file', { path, filePath }, 15000);
  }
}

export async function stageAll(path: string): Promise<void> {
  if (isTauri) {
    await invokeWithTimeout('stage_all', { path }, 25000);
  }
}

export async function unstageAll(path: string): Promise<void> {
  if (isTauri) {
    await invokeWithTimeout('unstage_all', { path }, 25000);
  }
}

export async function stageHunk(path: string, filePath: string, hunkIndex: number): Promise<void> {
  if (isTauri) {
    await invokeWithTimeout('stage_hunk', { path, filePath, hunkIndex }, 15000);
  }
}

export async function unstageHunk(path: string, filePath: string, hunkIndex: number): Promise<void> {
  if (isTauri) {
    await invokeWithTimeout('unstage_hunk', { path, filePath, hunkIndex }, 15000);
  }
}

export async function discardFileChanges(path: string, filePath: string): Promise<number> {
  if (isTauri) {
    return await invokeWithTimeout<number>('discard_file_changes', { path, filePath }, 20000);
  }
  return Date.now();
}

export async function discardAllChanges(path: string): Promise<number[]> {
  if (isTauri) {
    return await invokeWithTimeout<number[]>('discard_all_changes', { path }, 30000);
  }
  return [];
}

export async function listTrashSnapshots(path: string): Promise<TrashSnapshotItem[]> {
  if (isTauri) {
    return await invokeWithTimeout<TrashSnapshotItem[]>('list_trash_snapshots', { path }, 15000);
  }
  return getMockTrashSnapshots();
}

export async function restoreTrashSnapshot(path: string, snapshotId: number): Promise<void> {
  if (isTauri) {
    await invokeWithTimeout('restore_trash_snapshot', { path, snapshotId }, 20000);
  }
}

export async function restoreTrashBatch(path: string, batchId: string): Promise<number> {
  if (isTauri) {
    return await invokeWithTimeout<number>('restore_trash_batch', { path, batchId }, 30000);
  }
  return 0;
}

export async function deleteTrashSnapshot(snapshotId: number): Promise<void> {
  if (isTauri) {
    await invokeWithTimeout('delete_trash_snapshot', { snapshotId }, 15000);
  }
}

export async function getFileBlame(
  path: string,
  filePath: string,
  minLine?: number,
  maxLine?: number
): Promise<BlameHunkItem[]> {
  if (isTauri) {
    return await invokeWithTimeout<BlameHunkItem[]>(
      'get_file_blame',
      { path, filePath, minLine: minLine ?? null, maxLine: maxLine ?? null },
      25000
    );
  }
  return [];
}

export async function getFileHistory(
  path: string,
  filePath: string,
  limit?: number
): Promise<FileHistoryItem[]> {
  if (isTauri) {
    return await invokeWithTimeout<FileHistoryItem[]>('get_file_history', { path, filePath, limit }, 25000);
  }
  return [];
}
